File: tests/fixtures/impact-map/laravel/app/Reports/CompletionReport.php
<?php

namespace App\Reports;

use Illuminate\Support\Facades\DB;

class CompletionReport
{
    /** Completions per course for the admin dashboard. */
    public function countsByCourse(): array
    {
        return DB::select("
            SELECT course_id, COUNT(*) AS total
            FROM course_completions
            WHERE status = 'completed'
            GROUP BY course_id
        ");
    }
}

File: tests/fixtures/impact-map/mixed-architecture/backend/reports/completion_report.py
from sqlalchemy import text


def completions_by_course(connection) -> list[dict]:
    """Powers the admin completion dashboard."""
    rows = connection.execute(
        text(
            """
            SELECT course_id, COUNT(*) AS total
            FROM enrollments
            WHERE status = 'completed'
            GROUP BY course_id
            ORDER BY total DESC
            """
        )
    )
    return [dict(row._mapping) for row in rows]
